Tribler/Policies/credit_mining_util.py

```diff
@@ -0,0 +1,217 @@
+class TorrentManagerCM(TaskManager):
+    """
+    *Temporary* class to handle load torrent.
+
+    Adapted from TorrentManager in SearchGridManager
+    """
+    __single = None
+
+    def __init__(self, session):
+        super(TorrentManagerCM, self).__init__()
+        TorrentManagerCM.__single = self
+
+        self.session = session
+        self.torrent_db = self.session.open_dbhandler(NTFY_TORRENTS)
+        self.channelcast_db = self.session.open_dbhandler(NTFY_CHANNELCAST)
+        self.votecastdb = self.session.open_dbhandler(NTFY_VOTECAST)
+
+        self.dslist = []
+
+    def load_torrent(self, torrent, callback=None):
+        """
+        function to load torrent dictionary to torrent object.
+
+        From TorrentManager.loadTorrent in SearchGridManager
+        """
+
+        # session is quitting
+        if not (self.session and self.session.get_torrent_store() and self.session.lm.torrent_store):
+            return
+
+        if not isinstance(torrent, CollectedTorrent):
+            if torrent.torrent_id <= 0:
+                torrent_id = self.torrent_db.getTorrentID(torrent.infohash)
+                if torrent_id:
+                    torrent.update_torrent_id(torrent_id)
+
+            if not self.session.has_collected_torrent(torrent.infohash):
+                files = []
+                trackers = []
+
+                # see if we have most info in our tables
+                if isinstance(torrent, RemoteTorrent):
+                    torrent_id = self.torrent_db.getTorrentID(torrent.infohash)
+                else:
+                    torrent_id = torrent.torrent_id
+
+                trackers.extend(self.torrent_db.getTrackerListByTorrentID(torrent_id))
+
+                if 'DHT' in trackers:
+                    trackers.remove('DHT')
+                if 'no-DHT' in trackers:
+                    trackers.remove('no-DHT')
+
+                # replacement # self.downloadTorrentfileFromPeers(torrent, None)
+                if self.session.has_download(torrent.infohash):
+                    return False
+
+                if torrent.query_candidates is None or len(torrent.query_candidates) == 0:
+                    self.session.download_torrentfile(torrent.infohash, None, 0)
+                else:
+                    for candidate in torrent.query_candidates:
+                        self.session.download_torrentfile_from_peer(candidate, torrent.infohash, None, 0)
+
+                torrent = NotCollectedTorrent(torrent, files, trackers)
+
+            else:
+                tdef = TorrentDef.load_from_memory(self.session.get_collected_torrent(torrent.infohash))
+
+                if torrent.torrent_id <= 0:
+                    del torrent.torrent_id
+
+                torrent = CollectedTorrent(torrent, tdef)
+
+        # replacement # self.library_manager.addDownloadState(torrent)
+        for dl_state in self.dslist:
+            torrent.addDs(dl_state)
+
+        # return
+        if callback is not None:
```

Comment:
Use a deferred instead of alternatively calling a callback or returning the torrent.